kernel/hil/flash: add memory protection, impl/connect/test for OpenTitan

twilfredo opened the pull request:
### Pull Request Overview

This pull request adds a new `HIL` to interface flash memory protection (fmp). 

The hardware support for fmp may vary between vendors. For example, OpenTitan has ['advanced' control options](https://docs.opentitan.org/hw/ip/flash_ctrl/doc/#Reg_mp_region_cfg_0) but `nRF52840` only allows r/w control (see [section 6.3](https://infocenter.nordicsemi.com/pdf/nRF52840_PS_v1.0.pdf#%5B%7B%22num%22%3A153%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C85.039%2C123.823%2Cnull%5D)).

Since we don't want to limit the hardware functionality, this `HIL` adds a basic implementation which should work with hardware with basic functionality (i.e `nRF52840`) and an advanced `HIL` that extends the basic one (could be used by `OpenTitan` to allow full control of fmp.  An implementation example can be seen with the new additions to the OpenTitan flash controller in this PR. 

OpenTitan Specific Changes:

- Unit tests to test FMP functionality
- Driver implementation
- Add FMP rules to ROM/Kernel regions

Also, the addition of fmp to supported devices can address some of the concerns discussed in https://github.com/tock/tock/issues/3229

### Testing Strategy

Running the tests (new/old) on Verilator
```
�I00000 test_rom.c:81] Version:    earlgrey_silver_release_v5-5665-g217a0168b
Build Date: 2022-09-29, 14:35:47

I00001 test_rom.c:115] Test ROM complete, jumping to flash!
Unable to find otbn-rsa, disabling RSA support
OpenTitan initialisation complete. Entering main loop
[FLASH_CTRL] Test page read/write....
    [ok]
[FLASH_CTRL] Test page erase....
    [ok]
[FLASH_CTRL] Test memory protection basic....
    [ok]
[FLASH_CTRL] Test memory protection functionality....
    [ok]
trivial assertion... 
    [ok]
```

### TODO or Help Wanted

- RFC on HIL changes


### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

bradjc commented:
I have a couple questions:

1. Do we need a HIL? Do we expect that other platforms will be able to implement this HIL?

2. How do we reason about this interface in the context of an MPU/PMP? There is overlapping functionality, do we want to expose the additional complexity in Tock?

twilfredo commented:
> I have a couple questions:
> 
>     1. Do we need a HIL? Do we expect that other platforms will be able to implement this HIL?
> 

The primary motive behind this was to add support for OpenTitan flash mp features, so technically we *could* just expose an api for the driver (exclusive to OT) to implement these features, without the need for this HIL (similar to the lowrisc OTBN). However, looking at other hardware (like Nordic socs, nRF52840 for example) they have similar functionality present. I was thinking that It would be good for Tock to implement these features for devices where possible.  Since Tock does not have a way to currently do this, adding a HIL to govern flash mp would be a starting point.

>     2. How do we reason about this interface in the context of an MPU/PMP? There is overlapping functionality, do we want to expose the additional complexity in Tock?

I'm not too familiar with the MPU/PMP usage in Tock,

twilfredo commented:
> I have a couple questions:
> 
>     1. Do we need a HIL? Do we expect that other platforms will be able to implement this HIL?
> 
The primary motive behind this was to add support for OpenTitan flash mp features, so technically we *could* just expose an api for the driver (exclusive to OT) to implement these features, without the need for this HIL (similar to the lowrisc OTBN). However, looking at other hardware (like Nordic socs, nRF52840 for example) they have similar functionality present. I was thinking that It would be good for Tock to implement these features for devices where possible.  Since Tock does not have a way to currently do this, adding a HIL to govern flash mp would be a starting point.

>     2. How do we reason about this interface in the context of an MPU/PMP? There is overlapping functionality, do we want to expose the additional complexity in Tock?

I'm not super familiar with the MPU/PMP usage in Tock. Do we currently utilize these for flash memory protection, perhaps not all MPUs might support this(?).  @alistair23 maybe you can chime in on this abit?

hudson-ayers commented on kernel/src/hil/flash.rs:
```diff
@@ -142,3 +145,154 @@ pub trait Client<F: Flash> {
+/// Memory access control protection for flash. For hardware that supports
+/// flash memory protection, this can be used to implement the relevant functionality.
+///
+/// Requires the Flash interface to have been implemented as well.
```
In the context of the possibility for additional future traits as alternatives to the current `Flash` trait, do we need to require this?

twilfredo replied:
Wouldn't we want to make sure flash is implemented before flash memory protection is though?

hudson-ayers replied:
I don't think it could really hurt for something to implement this trait without implementing flash, it just wouldn't be useful. And if https://github.com/tock/tock/pull/2993 is merged, it is very possible there could be something that implements `BlockStorage` but not `Flash` which still might benefit from flash memory protection

hudson-ayers reviewed (commented):
My feeling is that a HIL only makes sense when we know of multiple chips with support, so the basic HIL makes sense but the advanced does not.

twilfredo commented:
> My feeling is that a HIL only makes sense when we know of multiple chips with support, so the basic HIL makes sense but the advanced does not.

Yea fair enough, would it make more sense to expose the `advanced` features as a part of the OpenTitan flash driver (new api ?)

bradjc commented:
We discussed this in the OT WG, and the conclusion was this should be an API in the lowrisc chip for now. There are two questions that are hard to answer right now as a HIL:

1. Will this HIL API work for a reasonable subset of chips that have this functionality?
2. How does this API intersect with the MPU trait?

By making it an API we can learn more about how to answer those questions without having to address them now.

twilfredo commented:
> 
that sounds good, I'll work on converting this into a chip api.

twilfredo commented:
Updates:
- Rebased
- Reverted HIL changes, and replaced with a chip specific public api for interfacing to `flash_mp`... similar to the `otbn` the implementation.
- Kept the `FlashMPError` code in the flash HIL, as this generic and could be used by other drivers.

twilfredo commented:
Ping!

alistair23 commented on boards/opentitan/src/main.rs:
```diff
@@ -220,6 +221,11 @@ impl KernelResources<earlgrey::chip::EarlGrey<'static, EarlGreyDefaultPeripheral
 }
 
+#[inline(always)]
+fn div_ceil(a: u32, b: u32) -> u32 {
```
It's worth adding a comment that we can remove this and use the Rust standard one at some point.

See: https://doc.rust-lang.org/std/primitive.i32.html#method.div_ceil and https://github.com/rust-lang/rust/issues/88581

alistair23 commented on boards/opentitan/src/main.rs:
```diff
@@ -457,6 +463,36 @@ unsafe fn setup() -> (
+
+    // Starting from page 0 (start of flash) to num_pages_occupied on `region 0`.
+    if let Err(e) =
+        peripherals
+            .flash_ctrl
+            .mp_set_region_perms(0, num_pages_occupied as usize, 0, &mp_cfg)
```
I feel like this API should match the MPU/PMP API. So in `main()` we can just pass in the same addresses and handle the conversion to pages in the flash controller. Especially as this isn't a generic HIL but an OT specific API

twilfredo replied:
sounds good, I'm thinking this:
```
peripherals
            .flash_ctrl
            .mp_set_region_perms(start_addr, end_addr, region_num, &mp_cfg)
```
and in `pub fn mp_set_region_perms()`, we check if the addresses are valid then convert to pages with a helper function to calculate the num pages between `start_addr & end_addr`.

Does that sound okay?

alistair23 replied:
That sounds good!

The only question left is do we need region number?

twilfredo replied:
cool! hmm... in this instance, we don't I suppose, since we are just setting up the single region. But later down the line could be useful if we wanted to have multiple isolated regions with varying levels of access control.

alistair23 replied:
Ok, might be best to provide a bit more info in the doc comments about what a region number is

alistair23 commented on chips/lowrisc/src/flash_ctrl.rs:
```diff
@@ -564,6 +610,187 @@ impl<'a> FlashCtrl<'a> {
+    /// * `num_pages` - Number of pages the configs are applied to (region size)
+    /// * `region_num`  - The configuration region number associated with this region.
+    ///                   This associates the specified permissions to a configuration region.
+    /// * `mp_perms` - Specifies the permissions to set
```
We should document the return types here as well

twilfredo commented:
Updates:

- The api now uses the following convention as discussed above
```
    pub fn mp_set_region_perms(
        &self,
        start_addr: usize,
        end_addr: usize,
        region_num: usize,
        mp_perms: &FlashMPConfig,
    )
```

- Added more comments for how regions work
- Added comments for return conditions
- Updated tests to match the new api
- Updated `main.rs` to match new api, now just passing in the addresses similar to MPU/PMP

twilfredo commented on boards/opentitan/src/main.rs:
```diff
@@ -457,6 +458,35 @@ unsafe fn setup() -> (
+    if let Err(e) =
+        peripherals
+            .flash_ctrl
+            .mp_set_region_perms(0x0, text_end_addr as usize, 0, &mp_cfg)
```
as for `_manifest` which in our case would be at 20000000h (start of bank 0 page 0), the flash registers relatively refer to this as address 0. Similarly for `_etext` which would be `20000000h + x`. 
![image](https://user-images.githubusercontent.com/36925352/200208726-ff845ade-3484-4b9f-845e-b3e5505db874.png)


We could pass in the actual addresses as `_manifest` and `_etext` and just subtract the offset...? But then we would have to keep an eye of this, in-case the base address changes.

twilfredo replied:
I do think passing `_manifest and _etext` is alot more cohesive though, should we do this and offset it in the driver?

alistair23 replied:
Yeah, I think we should pass the address and offset them in the driver

twilfredo commented on boards/opentitan/src/main.rs:
```diff
@@ -457,6 +458,35 @@ unsafe fn setup() -> (
+    let etext_addr = &_etext as *const u8 as u32;
+    let smanifest_addr = &_manifest as *const u8 as u32;
+    // Protect from ROM start to end of text
+    let text_end_addr = etext_addr.checked_sub(smanifest_addr).unwrap(); // Unwrap fail = addresses have updated in an unexpected way, this should be revised.
```
technically don't need this, I had it in there in-case the linker file in a weird way and this goes uncaught...but perhaps such a case wouldn't even be possible? where etext < smanifest ?

alistair23 replied:
You could add a linker assert to the board file, otherwise I think remove this. It adds a runtime overhead that isn't required

twilfredo commented:
Updates:
- Added a linker assert and removed the `unwrap()` code in `main.rs` for checking the address layout. 
- `mp_set_region_perms()` now takes the absolute addresses and offsets them in the driver.

bradjc commented on kernel/src/hil/flash.rs:
```diff
@@ -100,6 +100,9 @@ pub enum Error {
     /// An error occurred during the flash operation.
     FlashError,
+
+    /// A flash memory protection violation was detected
+    FlashMPError,
```
```suggestion
    /// A flash memory protection violation was detected.
    FlashMemoryProtectionError,
```

twilfredo replied:
done!

twilfredo commented:
Ping!

bradjc commented:
bors r+
